Import libraries

In [38]:
# Import required libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [39]:
# model building 
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error

Download data

In [40]:
url = "https://raw.githubusercontent.com/DavidCico/Boston-House-Prices-With-Regression-Machine-Learning-and-Keras-Deep-Learning/refs/heads/master/housing.csv"
df = pd.read_csv(url, sep='\s+', header=None)
df.columns = [
    'CRIM', 'ZN', 'INDUS', 'CHAS', 'NOX', 'RM', 'AGE', 'DIS',
    'RAD', 'TAX', 'PTRATIO', 'B', 'LSTAT', 'MEDV'
]

Explore and analyze data

In [41]:
X = df.drop("MEDV", axis=1)
y = df["MEDV"]

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

model = LinearRegression()
model.fit(X_train, y_train)

y_pred = model.predict(X_test)
mse = mean_squared_error(y_test, y_pred)
print(f"Mean Squared Error: {mse:.2f}")


Mean Squared Error: 24.29


Classification using Iris Dataset 

1st : Load additional libraries

In [42]:
from sklearn.preprocessing import LabelEncoder
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

2nd : Download data

In [43]:
url = "https://raw.githubusercontent.com/Apaulgithub/oibsip_taskno1/refs/heads/main/Iris.csv"
df = pd.read_csv(url)
df.head()

Id  SepalLengthCm  SepalWidthCm  PetalLengthCm  PetalWidthCm      Species
0   1            5.1           3.5            1.4           0.2  Iris-setosa
1   2            4.9           3.0            1.4           0.2  Iris-setosa
2   3            4.7           3.2            1.3           0.2  Iris-setosa
3   4            4.6           3.1            1.5           0.2  Iris-setosa
4   5            5.0           3.6            1.4           0.2  Iris-setosa

3rd : Data Preparation

In [49]:

# Separation of features and target
X = df.drop(["Id", "Species"], axis=1)  # نحذف 'Id' لأنها غير مفيدة للنموذج
y = df["Species"]

# Convert categories from words to numbers
from sklearn.preprocessing import LabelEncoder
le = LabelEncoder()
y = le.fit_transform(y)

# Split data into training and testing sets
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)


4th: Model Training & Logistic Regression

In [50]:
from sklearn.linear_model import LogisticRegression

# Create and train the model
model = LogisticRegression(max_iter=200)
model.fit(X_train, y_train)

LogisticRegression(max_iter=200)

5th: Model evaluation & accuracy score

In [51]:
from sklearn.metrics import accuracy_score

# Make predictions
y_pred = model.predict(X_test)

# Calculate accuracy
accuracy = accuracy_score(y_test, y_pred)
print(f"Accuracy: {accuracy * 100:.2f}%")

Accuracy: 100.00%


My report:
Supervised Learning (Regression & Classification)

Part 1: Boston Housing Price Prediction (Linear Regression)
So, I worked with the Boston Housing dataset to predict house prices. The file was kind of messy at first — no headers and weird spacing everywhere. I spent some time cleaning it up and added the column names manually using pandas.

The goal was to predict the MEDV column, which is the median home price. I split the data into 80% for training and 20% for testing.

After training a LinearRegression model, I got a Mean Squared Error of about 24.29. It’s not perfect, but it’s a decent start. I think with some more advanced methods, the results could definitely improve.

The biggest challenge for me was dealing with the dataset format — getting it loaded correctly wasn’t straightforward because of the missing headers and inconsistent spacing.

Part 2: Iris Flower Classification (Logistic Regression)
This part was easier. I used the Iris dataset to classify the flower species. I removed the Id column since it wasn’t useful, then encoded the species names into numbers. Like before, I split the data 80/20.

I trained a LogisticRegression model (with max_iter set to 200) and it scored 100% accuracy, which was pretty satisfying. The dataset was clean and balanced, so the model learned quickly.

Final Thoughts
Both models worked well overall. The linear regression gave reasonable predictions, and the logistic regression did perfectly on the classification task.

What I learned is that having clean and well-structured data is crucial — even simple models can perform surprisingly well if the data is good.

Next time, I want to try adding regularization or experimenting with more complex models like Decision Trees or Random Forests to see if I can get better results.

Stu   : Nada Abushawish 
StuNo : 2320225062